// ion-js/dist/amd/es6/IonBinary.js
/*!
 * Copyright 2012 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.NIBBLE_MASK = 0xf;
    exports.BYTE_MASK = 0xff;
    exports.TYPE_SHIFT = 4;
    exports.BYTE_SHIFT = 8;
    exports.LEN_MASK = 0xf;
    exports.LEN_VAR = 14;
    exports.LEN_NULL = 15;
    exports.TB_NULL = 0;
    exports.TB_BOOL = 1;
    exports.TB_INT = 2;
    exports.TB_NEG_INT = 3;
    exports.TB_FLOAT = 4;
    exports.TB_DECIMAL = 5;
    exports.TB_TIMESTAMP = 6;
    exports.TB_SYMBOL = 7;
    exports.TB_STRING = 8;
    exports.TB_CLOB = 9;
    exports.TB_BLOB = 10;
    exports.TB_LIST = 11;
    exports.TB_SEXP = 12;
    exports.TB_STRUCT = 13;
    exports.TB_ANNOTATION = 14;
});
